cpp/src/parquet/column_writer.cc

```diff
@@ -1490,7 +1490,8 @@ Status TypedColumnWriterImpl<DType>::WriteArrowDictionary(
     // TODO(wesm): If some dictionary values are unobserved, then the
```

Comment:
this comment should be deleted.  I also think you need to move this into WriteIndicesChunk above (I think you need to adjust your test to have batch size and page size small enough to generate multiple pages to catch this (I actually guess this is a different bug though).

Reply:
I've deleted the comment but I don't yet understand the point about WriteIndicesChunk.  I will need to look at this again later.  Is it a performance concern (don't make these extra allocations for the entire array all at once but make the extra allocations many times at the batch level to avoid total memory use) or is it a functional concern (maybe statistics are stored on a per-batch basis?)

Reply:
functional concern (unfortunately I think the performance would be worse since it will require extra allocations for each batch).  Statistics are stored at two levels rowgroup (column chunk) and page level.  The batching in done here and in other locations in the code to be able to get some level of vectorization but not make any individual page too large.  I might have traced the code incorrectly but the current location for updating statistics will only cover correct statistics at the row group level.  It seems like this is an orthogonal bug so you can maybe do it as a follow-up.

Reply:
Sorry for taking so long to get back to this but I dug into this a bit today and I can't find any page statistics.  Either in the code or the thrift definition.  Can you give me a pointer to the statistics objects you think might be incorrect.

There is a `PageEncodingStats` which has a `count` but that appears to be counting the number of pages and not counting the rows.

Reply:
https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L530 and https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L584 I thought I traced the code an we currently populate them but I could be mistaken.

Reply:
It will also be important if we ever implement [columindex writing](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L938)

Reply:
Sorry for the delay.  I've expanded the unit test to verify page level statistics.  It appears that we recompute the null count in `WriteIndicesChunk` and do so based on `def_levels` which appears to be calculated on a completely different code path (`MultipathLevelBuilderResult`).  So the concern about indices vs. dictionary does not apply and the null counts are correct.

I did notice we don't encode page level min/max stats at all.  I'm not sure if that is a bug or not (although, if so, I'd tackle that as a separate PR/JIRA).  So, if the unit test seems ok and the logic above seems valid then I think this is good.